File: examples/java/src/main/java/glide/examples/CompressionExample.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.examples;

import static glide.api.logging.Logger.Level.INFO;
import static glide.api.logging.Logger.log;

import glide.api.GlideClient;
import glide.api.logging.Logger;
import glide.api.models.configuration.CompressionBackend;
import glide.api.models.configuration.CompressionConfiguration;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.NodeAddress;
import java.util.Map;

/**
 * Demonstrates how to create a GLIDE client with transparent compression enabled.
 *
 * <p>When compression is enabled, the client automatically compresses values for SET-type commands
 * and decompresses values for GET-type commands, reducing bandwidth and storage usage.
 *
 * <p>Usage: Ensure a Valkey server is running on localhost:6379, then run this example.
 */
public class CompressionExample {

    public static void main(String[] args) throws Exception {
        Logger.setLoggerConfig(INFO);

        // Create compression configuration
        CompressionConfiguration compressionConfig =
                CompressionConfiguration.builder()
                        .enabled(true)
                        .backend(CompressionBackend.ZSTD) // or CompressionBackend.LZ4
                        .compressionLevel(3) // optional, backend-specific
                        .minCompressionSize(64) // only compress values >= 64 bytes
                        .build();

        // Build client configuration with compression
        GlideClientConfiguration config =
                GlideClientConfiguration.builder()
                        .address(NodeAddress.builder().host("localhost").port(6379).build())
                        .compressionConfiguration(compressionConfig)
                        .requestTimeout(5000)
                        .build();

        try (GlideClient client = GlideClient.createClient(config).get()) {
            // Write a large value — it will be compressed transparently
            String largeValue = "A".repeat(10) + "B".repeat(10) + "C".repeat(10);
            largeValue = largeValue.repeat(100); // ~3KB of compressible data

            client.set("compression_demo_key", largeValue).get();
            log(INFO, "example", "SET completed with compression enabled");

            // Read it back — decompression is automatic
            String retrieved = client.get("compression_demo_key").get();
            if (!retrieved.equals(largeValue)) {
                throw new RuntimeException("Value mismatch after decompression");
            }
            log(INFO, "example", "GET returned correct decompressed value");

            // Check compression statistics
            Map<String, String> stats = client.getStatistics();
            log(INFO, "example", "Compression statistics:");
            log(INFO, "example", "  Values compressed:   " + stats.get("total_values_compressed"));
            log(INFO, "example", "  Values decompressed: " + stats.get("total_values_decompressed"));
            log(INFO, "example", "  Original bytes:      " + stats.get("total_original_bytes"));
            log(INFO, "example", "  Compressed bytes:    " + stats.get("total_bytes_compressed"));
            log(INFO, "example", "  Skipped (too small): " + stats.get("compression_skipped_count"));

            // Verify compression actually happened and reduced size
            long compressed = Long.parseLong(stats.get("total_values_compressed"));
            long originalBytes = Long.parseLong(stats.get("total_original_bytes"));
            long compressedBytes = Long.parseLong(stats.get("total_bytes_compressed"));
            if (compressed <= 0) {
                throw new RuntimeException("Expected at least one value to be compressed");
            }
            if (compressedBytes >= originalBytes) {
                throw new RuntimeException(
                        "Expected compressed bytes ("
                                + compressedBytes
                                + ") < original bytes ("
                                + originalBytes
                                + ")");
            }
            log(
                    INFO,
                    "example",
                    "Compression ratio: "
                            + String.format("%.1f%%", (1.0 - (double) compressedBytes / originalBytes) * 100));

            // Cleanup
            client.del(new String[] {"compression_demo_key"}).get();
            log(INFO, "example", "Cleanup complete");
        }
    }
}

File: examples/csharp/CompressionExample.cs
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

/// <summary>
/// Compression Example
///
/// This example demonstrates how to create GLIDE clients with compression
/// configurations for both standalone and cluster modes.
///
/// Compression is transparent — values that meet the minimum size threshold
/// are automatically compressed before being sent to the server and
/// decompressed when retrieved. Compressed and uncompressed clients can
/// interoperate seamlessly.
/// </summary>

using Glide;

// ============================================================
// Example 1: Standalone client with ZSTD compression (default)
// ============================================================

var zstdCompression = new CompressionConfiguration(
    enabled: true,
    backend: CompressionBackend.Zstd,
    compressionLevel: 3,       // ZSTD default level
    minCompressionSize: 64     // Only compress values >= 64 bytes
);

var standaloneConfig = new GlideClientConfiguration(
    addresses: new[] { new NodeAddress("localhost", 6379) },
    compression: zstdCompression
);

Console.WriteLine("Standalone client configuration with ZSTD compression:");
Console.WriteLine($"  Enabled: {standaloneConfig.Compression!.Enabled}");
Console.WriteLine($"  Backend: {standaloneConfig.Compression.Backend}");
Console.WriteLine($"  Level: {standaloneConfig.Compression.CompressionLevel}");
Console.WriteLine($"  Min Size: {standaloneConfig.Compression.MinCompressionSize} bytes");
Console.WriteLine();

// ============================================================
// Example 2: Cluster client with LZ4 compression
// ============================================================

var lz4Compression = new CompressionConfiguration(
    enabled: true,
    backend: CompressionBackend.Lz4,
    compressionLevel: 0,       // LZ4 default level
    minCompressionSize: 128    // Only compress values >= 128 bytes
);

var clusterConfig = new GlideClusterClientConfiguration(
    addresses: new[]
    {
        new NodeAddress("node1.example.com", 6379),
        new NodeAddress("node2.example.com", 6379),
        new NodeAddress("node3.example.com", 6379),
    },
    compression: lz4Compression
);

Console.WriteLine("Cluster client configuration with LZ4 compression:");
Console.WriteLine($"  Enabled: {clusterConfig.Compression!.Enabled}");
Console.WriteLine($"  Backend: {clusterConfig.Compression.Backend}");
Console.WriteLine($"  Level: {clusterConfig.Compression.CompressionLevel}");
Console.WriteLine($"  Min Size: {clusterConfig.Compression.MinCompressionSize} bytes");
Console.WriteLine();

// ============================================================
// Example 3: Compression with backend default level
// ============================================================

var defaultLevelCompression = new CompressionConfiguration(
    enabled: true,
    backend: CompressionBackend.Zstd
    // compressionLevel omitted — uses backend default (3 for ZSTD)
    // minCompressionSize omitted — uses default (64 bytes)
);

Console.WriteLine("Compression with backend default level:");
Console.WriteLine($"  Level: {(defaultLevelCompression.CompressionLevel?.ToString() ?? "backend default")}");
Console.WriteLine($"  Min Size: {defaultLevelCompression.MinCompressionSize} bytes");
Console.WriteLine();

// ============================================================
// Example 4: Client without compression (default behavior)
// ============================================================

var noCompressionConfig = new GlideClientConfiguration(
    addresses: new[] { new NodeAddress("localhost", 6379) }
);

Console.WriteLine($"Client without compression: Compression is {(noCompressionConfig.Compression == null ? "not configured" : "configured")}");
Console.WriteLine();

// ============================================================
// Example 5: Setting compression after client config creation
// ============================================================

var config = new GlideClientConfiguration(
    addresses: new[] { new NodeAddress("localhost", 6379) }
);

// Compression can be set after initial configuration
config.Compression = new CompressionConfiguration(
    enabled: true,
    backend: CompressionBackend.Zstd,
    compressionLevel: 5,
    minCompressionSize: 256
);

Console.WriteLine("Compression set after config creation:");
Console.WriteLine($"  Enabled: {config.Compression.Enabled}");
Console.WriteLine($"  Level: {config.Compression.CompressionLevel}");
Console.WriteLine($"  Min Size: {config.Compression.MinCompressionSize} bytes");
Console.WriteLine();

// ============================================================
// Example 6: Protobuf serialization for FFI
// ============================================================

var bytes = ConnectionRequestSerializer.Serialize(standaloneConfig, clusterMode: false);
Console.WriteLine($"Serialized ConnectionRequest: {bytes.Length} bytes");
Console.WriteLine();

// ============================================================
// Example 7: Statistics usage
// ============================================================

// In a real application, statistics would be retrieved from the FFI layer.
// This demonstrates the Statistics class structure.
var stats = new Statistics
{
    TotalConnections = 3,
    TotalClients = 1,
    TotalValuesCompressed = 150,
    TotalValuesDecompressed = 120,
    TotalOriginalBytes = 500000,
    TotalBytesCompressed = 125000,
    TotalBytesDecompressed = 500000,
    CompressionSkippedCount = 30,
};

Console.WriteLine("Compression Statistics:");
Console.WriteLine($"  Values compressed: {stats.TotalValuesCompressed}");
Console.WriteLine($"  Values decompressed: {stats.TotalValuesDecompressed}");
Console.WriteLine($"  Original bytes: {stats.TotalOriginalBytes}");
Console.WriteLine($"  Compressed bytes: {stats.TotalBytesCompressed}");
Console.WriteLine($"  Compression ratio: {(double)stats.TotalOriginalBytes / stats.TotalBytesCompressed:F2}x");
Console.WriteLine($"  Skipped: {stats.CompressionSkippedCount}");
Console.WriteLine();

var dict = stats.ToDictionary();
Console.WriteLine($"Statistics as dictionary ({dict.Count} entries):");
foreach (var (key, value) in dict)
{
    Console.WriteLine($"  {key}: {value}");
}

Console.WriteLine();
Console.WriteLine("Compression example completed successfully.");
